[browser] high level documentation of WASM features

pavelsavara opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/80098
Related https://github.com/dotnet/runtime/issues/76722

pavelsavara committed:
wip

pavelsavara committed:
whitespace

pavelsavara committed:
wip

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+
+dotnet runtime for wasm is CSP compliant starting from .Net 8, except legacy JS interop methods.
+
+In order to enable it, please set HTTP headers similar to `Content-Security-Policy: default-src 'self' 'wasm-unsafe-eval'`
```
We could also specify that wasm-unsafe-eval is not needed if you disable the jiterpreter

pavelsavara replied:
I think until could could ES6 `import` wasm module, the `wasm-unsafe-eval` is always necessary. Because we just compile some bytes.

See https://github.com/WebAssembly/content-security-policy/blob/main/proposals/CSP.md#the-wasm-unsafe-eval-source-directive

I was not sure, so I tested it even with `WebAssembly.instantiateStreaming` and still I get `CompileError: WebAssembly.instantiateStreaming(): Refused to compile or instantiate WebAssembly module because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self'"`

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
-We also use v8 engine version 11 or higher to run some of the tests. The engine is lacking many APIs and features.
\ No newline at end of file
+
+We also use v8 engine version 11 or higher to run some of the tests. The engine is lacking many APIs and features.
```
In this case it's not "the engine" lacking many APIs and features but instead the d8 shell, which is what jsvu is installing. So something like this might be better:

`We also use the d8 command-line shell, version 11 or higher, to run some of the tests. This shell lacks most browser APIs and features.`

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+
+# Mobile phones
+
+Recent mobile phones have browser integration which could be upgraded separately from the operating system.
```
Something like `Recent mobile phones distribute their browser as an application that can be upgraded separately from the operating system.` might be clearer here.

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+
+Recent mobile phones have browser integration which could be upgraded separately from the operating system.
+
+Also note that all browsers on iOS are just wrapper for Safari engine and so it's limited by Safari features and version, rather than wrapper version and brand.
```
I suggest `Note that all browsers on iOS and iPadOS are required to use the Safari browser engine, so their level of support for WASM features depends on the version of Safari installed on the device.`

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+
+Also note that all browsers on iOS are just wrapper for Safari engine and so it's limited by Safari features and version, rather than wrapper version and brand.
+
+Mobile phones usually have limited resources. Memory and download speed are major concern.
```
I suggest `Mobile browsers typically have strict limits on the amount of memory they can use, and many users are on slow internet connections. A WebAssembly application that works well on desktop PCs may take minutes to download or run out of memory before it is able to start.`

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+## WASM linear memory
+
+You can override initial size of the WASM linear memory by `<EmccInitialHeapSize>16777216</EmccInitialHeapSize>`.
+This requires that you have [wasm-tools workload](#wasm-tools-workload) installed.
```
We could also explain briefly here why you might choose to do this, i.e. `Setting an initial size based on how much memory your application typically uses will reduce the number of times the heap needs to grow, which may enable it to run on devices with lower memory`. I assume that's the reason for doing this, anyway.

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+
+## JITerpreter
+
+Is browser specific JIT compiler which optimizes small fragments of code which are otherwise interpreted by dotnet mono interpreter. It's enabled by default.
```
We could state here that it will increase memory usage, and mention that the jiterpreter documentation (with a link to it) explains how to disable it.

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+This requires that you have [wasm-tools workload](#wasm-tools-workload) installed.
+
+It will compile managed code as native WASM instructions and include them in the `dotnet.native.wasm` file.
+AOT compiled code is running faster but it will significantly increase size of the file and the download time.
```
Something like `AOT compilation greatly improves application performance but will increase the size of the application, resulting in longer downloads and slower startup.`

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+
+## IL trimming
+
+You can trim size of the managed code in the assemblies by `<PublishTrimmed>true</PublishTrimmed>`.
```
It's probably worth explaining in more detail what IL trimming does and why you would want to use it, like
`Trimming will remove unused code from your application, which reduces download time and memory usage. When AOT compilation is in use, trimming will also reduce the amount of time spent to compile the application.`

kg replied:
We may also want to warn that `Some applications will break if trimming is used without further configuration due to the trimmer not knowing which code is used` and link to the documentation on trimming so users can configure it.

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+## C code or native linked libraries
+
+You can enable native rebuild by `<WasmBuildNative>true</WasmBuildNative>`.
+This requires that you have [wasm-tools workload](#wasm-tools-workload) installed.
```
We should consider explaining what native rebuild does and why you would use it, like `native rebuild will cause the entire .NET runtime to be re-built alongside your application, which allows you to link additional libraries into the binary or change compiler configuration flags.`

kg commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,125 @@
+
+## wasm-tools workload
+
+Is set of compiler tools which allow you to optimize your wasm application.
```
Something like `The optional wasm-tools workload contains all of the tools and libraries necessary to perform native rebuild or AOT compilation` might be better here

pavelsavara committed:
Merge branch 'main' into browser_features

pavelsavara committed:
feedback from @kg

pavelsavara commented:
many thanks @kg!

pavelsavara committed:
Update features.md

pavelsavara committed:
Update features.md

pavelsavara committed:
Update features.md

pavelsavara committed:
Update features.md

ilonatommy commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,270 @@
+
+If your application doesn't need to work with locatization, you can reduce download size by `<InvariantCulture>true</InvariantCulture>`.
+
+If your application needs to work with locatization but you don't require processing speed for processing of large texts, you can reduce download size by `<HybridGlobalization>true</HybridGlobalization>`.
```
Link this one https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-hybrid-mode.md

ilonatommy commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,270 @@
+If your application needs to work with locatization but you don't require processing speed for processing of large texts, you can reduce download size by `<HybridGlobalization>true</HybridGlobalization>`.
+It will use browser APIs instead part of ICU database. This feature is still in development.
+
+If your application needs to work with specific subset of locales, you can create your own ICU database.
```
Here we are missing info about sharding and general logic of ICU files (choosing based on culture etc.) that is included in https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-icu-wasm.md, so I would just redirect to that link. "To learn how Globalization works when Invariant and Hybrid modes are off, see: [doc]."

ilonatommy reviewed (approved):
I read all, looks good.

ilonatommy commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,270 @@
+
+Browsers don't offer full set APIs for working with localization and so we have to bring the data and the logic as part of the application.
+
+If your application doesn't need to work with locatization, you can reduce download size by `<InvariantCulture>true</InvariantCulture>`.
```
We could link the general doc about Invariant mode here: https://github.com/dotnet/runtime/blob/main/docs/design/features/globalization-invariant-mode.md

pavelsavara committed:
ICU

pavelsavara committed:
fetch api

pavelsavara committed:
WS API

pavelsavara committed:
embedding

pavelsavara committed:
Update features.md

kg committed:
Checkpoint copyediting and revision

kg committed:
Finished copyediting/revision pass

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+# Downloaded assets
+
+## Project folder structure
```
```suggestion
## Project folder structure for the `wasmbrowser` template
```

pavelsavara replied:
That structure is now same for all targets

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+## Project folder structure
+
+Described as relative to simple application project.
+
```
```suggestion

### Source directory

```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+Described as relative to simple application project.
+
+- `./wwwroot` is optional project folder, into which you can place files which should be also deployed to the web server.
+- `./bin/Release/net8.0/browser-wasm/AppBundle` folder which should be hosted by the HTTP server.
```
```suggestion
### Build output directories

The following shows the structure for a Release build, but except the name in the various paths, the rest is applicable for a `Debug` build too.

- `./bin/Release/net8.0/browser-wasm/AppBundle` folder which should be hosted by the HTTP server.
```

pavelsavara replied:
I see, splitting it with headers makes sense

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- `./bin/Release/net8.0/browser-wasm/AppBundle` folder which should be hosted by the HTTP server.
+- `./bin/Release/net8.0/browser-wasm/AppBundle/index.html` - the page which is hosting the application.
+- `./bin/Release/net8.0/browser-wasm/AppBundle/main.js` - typically the main JavaScript entry point, it will `import { dotnet } from './_framework/dotnet.js'` and `await dotnet.run();`.
+- `./bin/Release/net8.0/browser-wasm/AppBundle/_framework` - contains all the assets of the runtime and the managed application.
+
+You can flatten the `_framework` folder away by `<WasmRuntimeAssetsLocation>./</WasmRuntimeAssetsLocation>` in the project file.
+
+## `_framework` folder structure
```
```suggestion

#### `./bin/Release/net8.0/browser-wasm/AppBundle/_framework` folder structure

- `./bin/Release/net8.0/browser-wasm/AppBundle/_framework` - contains all the assets of the runtime and the managed application.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- `*.wasm` - are .NET assemblies wrapped as .wasm files in WebCIL forma for better compatibility with antivirus solutions.
+- `dotnet.js.map` - is source map file, for easier debugging of the runtime code. It's not included in published apps.
+- `dotnet.native.js.symbols` - are debug symbols which help to put C runtime method names back to the .wasm stack traces. You could enable it by `<WasmEmitSymbolMap>true</WasmEmitSymbolMap>`.
+
```
```suggestion

Note: You can flatten the `_framework` folder away by `<WasmRuntimeAssetsLocation>./</WasmRuntimeAssetsLocation>` in the project file.


```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+## `_framework` folder structure
+- `dotnet.js` - is the main entrypoint with the [JavaScript API](#JavaScript-API). It will load the rest of the runtime.
+- `dotnet.native.js` - is posix emulation layer by [emscripten](https://github.com/emscripten-core/emscripten) project
```
```suggestion
- `dotnet.native.js` - is a posix emulation layer provided by the [Emscripten](https://github.com/emscripten-core/emscripten) project
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- `dotnet.js` - is the main entrypoint with the [JavaScript API](#JavaScript-API). It will load the rest of the runtime.
+- `dotnet.native.js` - is posix emulation layer by [emscripten](https://github.com/emscripten-core/emscripten) project
+- `dotnet.runtime.js` - is integration of the dotnet with the browser
+- `blazor.boot.json` - contains list of all other assets and their integrity hash and also various configuration flags.
```
```suggestion
- `blazor.boot.json` - contains list of all other assets and their integrity hash and also various configuration flags. Note: even though this file has `blazor` in its name, it is completely independent of `Blazor`.
```

pavelsavara replied:
At this point I'm not sure I love that we have that file name. I would not elaborate it in the doc.

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- `dotnet.native.js` - is posix emulation layer by [emscripten](https://github.com/emscripten-core/emscripten) project
+- `dotnet.runtime.js` - is integration of the dotnet with the browser
+- `blazor.boot.json` - contains list of all other assets and their integrity hash and also various configuration flags.
+- `dotnet.native.wasm` - is the compiled binaries of the dotnet (Mono) runtime.
```
```suggestion
- `dotnet.native.wasm` - is the compiled binary of the dotnet (Mono) runtime.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- `blazor.boot.json` - contains list of all other assets and their integrity hash and also various configuration flags.
+- `dotnet.native.wasm` - is the compiled binaries of the dotnet (Mono) runtime.
+- `System.Private.CoreLib.wasm`
+- `*.wasm` - are .NET assemblies wrapped as .wasm files in WebCIL forma for better compatibility with antivirus solutions.
```
```suggestion
- `*.wasm` - are .NET assemblies wrapped as `.wasm` files in `WebCIL` format for better compatibility with antivirus solutions.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- `dotnet.native.wasm` - is the compiled binaries of the dotnet (Mono) runtime.
+- `System.Private.CoreLib.wasm`
+- `*.wasm` - are .NET assemblies wrapped as .wasm files in WebCIL forma for better compatibility with antivirus solutions.
+- `dotnet.js.map` - is source map file, for easier debugging of the runtime code. It's not included in published apps.
```
```suggestion
- `dotnet.js.map` - is a source map file, for easier debugging of the runtime code. It's not included in the published apps.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- `System.Private.CoreLib.wasm`
+- `*.wasm` - are .NET assemblies wrapped as .wasm files in WebCIL forma for better compatibility with antivirus solutions.
+- `dotnet.js.map` - is source map file, for easier debugging of the runtime code. It's not included in published apps.
+- `dotnet.native.js.symbols` - are debug symbols which help to put C runtime method names back to the .wasm stack traces. You could enable it by `<WasmEmitSymbolMap>true</WasmEmitSymbolMap>`.
```
```suggestion
- `dotnet.native.js.symbols` - are debug symbols which help to put `C` runtime method names back into the `.wasm` stack traces. You can enable it by `<WasmEmitSymbolMap>true</WasmEmitSymbolMap>`.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+_TODO_ discuss cache busting strategies.
+
+In order to make sure that the application resources are consistent with each other, dotnet runtime will use `integrity` hash for all downloaded resources.
```
```suggestion
In order to make sure that the application resources are consistent with each other, the dotnet runtime will use `integrity` hashes for all the downloaded resources.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+## MIME types
+
+Are `Content-Type` HTTP headers which tell the browser about the type of downloaded asset. They are necessary for correct and fast processing by the browser, but also by various caches and proxies.
```
```suggestion
Are `Content-Type` HTTP headers which tell the browser about the type of the downloaded asset. They are necessary for correct and fast processing by the browser, but also needed by various caches and proxies.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+## Compression
+
+Modern browsers are able to unpack files compressed for the transport. We recommend brotli compression for best results.
```
```suggestion
Modern browsers are able to unpack files compressed for the transport. We recommend `brotli` compression for best results.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+## Content security policy
+
+dotnet runtime for wasm is CSP compliant starting from .Net 8, except legacy JS interop methods.
```
```suggestion
dotnet runtime for wasm is CSP compliant starting with .Net 8, except legacy JS interop methods.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+Browsers don't offer full set APIs for working with localization and so we have to bring the data and the logic as part of the application.
+
+In order to make downloades smaller, runtime will detect the locale of the browser and load just shard of the ICU database. You can also create and configure your own custom shards.
```
```suggestion
In order to make downloads smaller, the runtime will detect the locale of the browser and load just a shard of the ICU database, which would only contain a few locales. You can also create and configure your own custom shards.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+For more details see [globalization-icu-wasm.md](../../../docs/design/features/globalization-icu-wasm.md)
+
+If your application doesn't need to work with locatization, you can reduce download size by `<InvariantCulture>true</InvariantCulture>`.
```
Did you mean `InvariantGlobalization`?
```suggestion
If your application doesn't need to work with localization, you can reduce the download size by `<InvariantGlobalization>true</InvariantGlobalization>`.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+For more details see [globalization-invariant-mode.md](../../../docs/design/features/globalization-invariant-mode.md)
+
+If your application needs to work with locatization but you don't require processing speed for working with large texts, you can reduce download size by `<HybridGlobalization>true</HybridGlobalization>`.
+It will use browser APIs instead part of ICU database. This feature is still in development.
```
```suggestion
If your application needs to work with localization but you don't require fast processing speed for working with large texts, you can reduce the download size by `<HybridGlobalization>true</HybridGlobalization>`. It will use browser APIs instead parts of the ICU database. This feature is still in development.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+Browsers don't offer API for working with time zone database and so we have to bring the time zone data as part of the application.
+
+If your application doesn't need to work with time zones, you can reduce download size by `<InvariantTimezone>true</InvariantTimezone>`.
```
```suggestion
If your application doesn't need to work with time zones, you can reduce the download size by `<InvariantTimezone>true</InvariantTimezone>`.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+You can bundle the `dotnet.js` ES6 module with the rest of your JavaScript application.
+
+The other assets and JS modules of the dotnet are loaded via dynamic `import` or via `fetch` APIs. They are not ready to be bundled in Net8.
+We consider that the dotnet application is usually large as is and giving the browser chance to start compiling it in parallel with other downloads is better.
```
```suggestion
We consider that the dotnet application is usually large as-is and giving the browser a chance to start compiling it in parallel with other downloads is better.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+The other assets and JS modules of the dotnet are loaded via dynamic `import` or via `fetch` APIs. They are not ready to be bundled in Net8.
+We consider that the dotnet application is usually large as is and giving the browser chance to start compiling it in parallel with other downloads is better.
+
+We would like to [hear from the community](https://github.com/dotnet/runtime/issues/86162) more about the use-cases when it would be benefitial.
```
```suggestion
We would like to [hear from the community](https://github.com/dotnet/runtime/issues/86162) more about the use-cases where it would be beneficial.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- integration with the browser JavaScript APIs, for example HTTP and WebSocket client
+- application code
+
+All of the mentioned code and data need to be downloaded and loaded into the browser memory during the dotnet startup sequence.
```
```suggestion
All of the mentioned code and data needs to be downloaded and loaded into the browser memory during the dotnet startup sequence.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+- application code
+
+All of the mentioned code and data need to be downloaded and loaded into the browser memory during the dotnet startup sequence.
+Browser itself will run JIT compilation of the WASM and JS code, which consumes memory and CPU cycles too.
```
```suggestion
The browser itself will run the JIT compilation of the WASM and JS code, which consumes memory and CPU cycles.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+## Trade-offs compared to native JavaScript applications
+
+There is trade-off between download size, application performance and complexity of application maintanance.
```
```suggestion
There is a trade-off between download size, application performance and complexity of application maintenance.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+## Trade-offs compared to native JavaScript applications
+
+There is trade-off between download size, application performance and complexity of application maintanance.
+If your business logic is very complex, changes often or you already have existing C# codebase and skillset, running the same code dotnet on wasm is probably the right choice.
+If your application is simple and you have JavaScript skills on your team, it may be better if you re/write your logic in Web native technologies like HTML/JavaScript or typescript/webpack stack.
```
```suggestion
- If your business logic is very complex, changes often or you already have existing C# codebase and skillset, then running the same code with dotnet on wasm is probably the right choice.
- If your application is simple and you have JavaScript skills on your team, it may be better if you re/write your logic in Web native technologies like HTML/JavaScript or typescript/webpack stack.
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+# wasm-tools workload
+
+The wasm-tools workload contains all of the tools and libraries necessary to perform native rebuild or AOT compilation and other optimizations of your application.
+
+Although it's optional for Blazor, we strongly recommend using it!
```
```suggestion
The wasm-tools workload contains all of the tools and libraries necessary to perform native rebuild or AOT compilation and other optimizations of your application. Although it's optional for Blazor, we strongly recommend using it!
```

radical commented on src/mono/wasm/features.md:
```diff
@@ -2,36 +2,311 @@
+
+You can install it by running `dotnet workload install wasm-tools` on your command line.
+
+You can also install `dotnet workload install wasm-experimental`.
```
Something about what's in this?

pavelsavara replied:
Is that the correct name for the workload with MT and templates ?

radical reviewed (changes requested):
Very comprehensive! Great job!!

These are my comments on first reading, feel free to accept/reject on a case by case basis.

Probably needs another look to format (maybe like bullets instead of new lines, or some things in small paragraphs) better, but super nit.

pavelsavara committed:
Merge pull request #5 from kg/pavel-docs

Copyediting and revision

pavelsavara committed:
more

pavelsavara commented:
@radical - your feedback came just after Kate did lot of great editing, which I merged. I went over your comments and I think she resolved most of them. I addressed some of them too. Thanks!

I applied your feedback about headers to project folders.

@kg thanks for the update! The only part I reverted a bit was "Choosing the right platform target" because I want to be quite explicit about the comparison. This not marketing doc after all.

I also added section about `pre-fetch` and section about `ETag` and `no-cache`.
I also added TOC.

I think we are close to good enough :)

Final comments ?

pavelsavara committed:
whitespace

lewing commented on src/mono/wasm/features.md:
```diff
@@ -1,37 +1,305 @@
+This feature only works if you have the [wasm-tools workload](#wasm-tools-workload) installed.
+
+### IL trimming
+Trimming will remove unused code from your application, which reduces application startup time and memory usage. Trimming also reduces the amount of time spent during AOT compilation if it is in use. To enable trimming of managed code, use `<PublishTrimmed>true</PublishTrimmed>` and `<EnableAggressiveTrimming>true</EnableAggressiveTrimming>`.
```
```suggestion
Trimming will remove unused code from your application, which reduces application startup time and memory usage. Trimming also reduces the amount of time spent during AOT compilation if it is in use. To enable trimming of managed code, use `<PublishTrimmed>true</PublishTrimmed>` and `<TrimMode>full</TrimMode>`.
```

EnableAggressiveTrimming is really a setting for use inside dotnet/runtime not for general consumption

pavelsavara committed:
Update src/mono/wasm/features.md

Co-authored-by: Larry Ewing <lewing@microsoft.com>

pavelsavara committed:
more
